dbms/src/Storages/StorageDisaggregated.cpp

```diff
@@ -0,0 +1,292 @@
+// Copyright 2022 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <DataStreams/TiRemoteBlockInputStream.h>
+#include <Flash/Coprocessor/InterpreterUtils.h>
+#include <Storages/StorageDisaggregated.h>
+#include <Storages/Transaction/TMTContext.h>
+
+namespace DB
+{
+const String StorageDisaggregated::ExecIDPrefixForTiFlashStorageSender = "exec_id_disaggregated_tiflash_storage_sender";
+
+BlockInputStreams StorageDisaggregated::read(
+    const Names &,
+    const SelectQueryInfo &,
+    const Context &,
+    QueryProcessingStage::Enum &,
+    size_t,
+    unsigned num_streams)
+{
+    auto remote_table_ranges = buildRemoteTableRanges();
+
+    auto batch_cop_tasks = buildBatchCopTasks(remote_table_ranges);
```

Comment:
What happens when `batch_cop_tasks .empty()`?

Reply:
Would we return a `NullBlockInputStream` directly like `DAGStorageInterpreter`.

Reply:
So we just need a `RUNTIME_CHECK`